### ISSUE TITLE
no-unused-vars false positive after upgrading to v3.0.0

### ISSUE THREAD

**Issue by gustavohenke** [eslint/eslint#6576]
**What version of ESLint are you using?**
3.0.0

**What parser (default, Babel-ESLint, etc.) are you using?**
Default.

**Please show your full configuration:**
```
{
	"root": true,
	"env": {
		"browser": true,
		"es6": true,
		"amd": true,
		"jquery": true
	},
	"globals": {
		"_": false,
		"angular": false,
		"CKEDITOR": false,
		"google": false
	},
	"parserOptions": {
		"ecmaVersion": 6
	},
	"rules": {
		// Possible errors
		"no-dupe-keys":                  2,
		"no-empty":                      1,
		"no-ex-assign":                  1,
		"no-irregular-whitespace":       2,

		// Best practices
		"curly":                       [ 2, "all" ],
		"dot-notation":                [ 2, {
			"allowKeywords": true
		}],
		"dot-location":                [ 2, "property" ],
		"eqeqeq":                      [ 2, "allow-null" ],
		"guard-for-in":                  1,
		"no-else-return":                1,
		"no-labels":                     2,
		"no-extend-native":              2,
		"no-fallthrough":                0,
		"no-lone-blocks":                1,
		"no-loop-func":                  2,
		"no-multi-spaces":               0,
		"no-multi-str":                  2,
		"no-native-reassign":            2,
		"no-param-reassign":             0,
		"no-return-assign":              0,
		"no-unused-expressions":         0,
		"radix":                         1,
		"yoda":                          2,

		// Variables
		"no-undef":                      2,
		"no-unused-vars":                2,
		"no-use-before-define":        [ 2, "nofunc" ],

		// Stylistic issues
		"array-bracket-spacing":       [ 2, "always", {
			"objectsInArrays": false
		}],
		"brace-style":                 [ 2, "1tbs", {
			"allowSingleLine": true
		}],
		"camelcase":                   [ 2, {
			"properties": "never"
		}],
		"comma-spacing":                 2,
		"comma-style":                   2,
		"computed-property-spacing":   [ 2, "always" ],
		"eol-last":                      0,
		"indent":                      [ 2, "tab", {
			"SwitchCase": 1
		}],
		"key-spacing":                   0,
		"lines-around-comment":        [ 2, {
			"allowBlockStart": true
		}],
		"max-nested-callbacks":        [ 1, 5 ],
		"new-parens":                    2,
		"newline-after-var":             0,
		"no-spaced-func":                0,
		"object-curly-spacing":        [ 2, "always" ],
		"one-var":                     [ 2, {
			"initialized":   "never"
		}],
		"quotes":                      [ 2, "double" ],
		"keyword-spacing":             [ 2, {
			"before": true,
			"after": true
		}],
		"space-in-parens":               0,
		"space-infix-ops":               2,
		"space-unary-ops":             [ 2, {
			"words": true,
			"nonwords": false
		}],
		"spaced-comment":              [ 2, "always" ],

		// Legacy
		"max-len":                     [ 2, 150 ]
	}
}
```

**What did you do? Please include the actual source code causing the issue.**
```js
var unregisterFooWatcher;
// ...
unregisterFooWatcher = $scope.$watch( "foo", function( foo ) {
	// ...some code..
	unregisterFooWatcher();
});
```

**What did you expect to happen?**
Code linted just fine.

**What actually happened?**
```
/path/to/file.js
  29:8  error  'unregisterFooWatcher' is defined but never used  no-unused-vars
```

Seems to be caused by modifications of #6348 and #6535.

**Comment by gustavohenke** [comment]
To be clear, our builds started failing a few hours ago because we run `npm install -g eslint` (no version specified).

For now, I'll just specify version 2..

**Comment by nzakas** [comment]
cc @mysticatea

**Comment by mysticatea** [comment]
Ugh, I'm sorry.
I will fix it.

**Comment by doberkofler** [comment]
@mysticatea Is this the same false positive I found in eslint 3.0.0 ?
```
/*eslint no-unused-vars:2*/
let _timer;
function f() {
	_timer = setTimeout(function () {}, _timer ? 100 : 0);
}
f();
```
reports:
```
3:5  warning  '_timer' is defined but never used  no-unused-vars
```

**Comment by yphoenix** [comment]
This looks like the same bug I'm seeing with:

`function foo(){
    'use strict';

    var ref;

    ref = setInterval(
        function(){
            clearInterval(ref);
        }, 10);
}`

As soon as I upgraded to 3.0.0
